pkg/agent/util/syscall/syscall_windows.go

```diff
@@ -368,7 +399,16 @@ func (n *netIO) ListIPForwardRows(family uint16) ([]MibIPForwardRow, error) {
 	if err != nil {
 		return nil, os.NewSyscallError("iphlpapi.GetIpForwardTable", err)
 	}
-	return unsafe.Slice(&table.Table[0], table.NumEntries), nil
```

Comment:
As far as I know the only issue with the previous code is that we were not making a copy of the slice prior to returning, while at the same time freeing the underlying memory by calling `n.freeMibTable(unsafe.Pointer(table))`.

The Wireguard Windows code has a simpler implementation of this, which still uses `unsafe.Slice`. It achieves correctness by making a proper copy of the slice: https://github.com/WireGuard/wireguard-windows/blob/e70799b1440690e7d4140bffc7c73baf903c7b54/tunnel/winipcfg/winipcfg.go#L151

That being said, if this new version has already been tested thoroughly, there is probably no string reason to change it.

Reply:
@antoninbas , we used to think about the solution as what Wiregurad is taking, a concern is some fields in MibIPForwardRow are using type `[]byte`, which still exists the risk with a shadow copy if the memory resource is limited (e.g., Windows OS re-allocate the memory to other objects after the `MibIPForwardTable` is free). So we finally choose the solution to deep copy the objects before free the memory. The latest code is verified on the problematic cluster.

Reply:
Do you mean `[N]byte` types like the `[26]byte` field in `RawSockAddrInet` (`[]byte` would not be possible here AFAIK)? These should be no different than integer types, they should be inlined in the struct with no indirection.

The code before this change was wrong for obvious reasons. I don't think the Wireguard code is wrong. But if you feel like manual copy is the best approach and this has been well-tested, as I wrote before we can merge it as is.

Reply:
The fields in MibIPForwardRow is array not slice. It will be deepcopied automatically.
And the current code is inconsistent: it makes a manual deepcopy for RawSockAddrInet but not AddressPrefix which also has a RawSockAddrInet.

Reply:
We also called `AddressPrefix.deepcopy()`, there is also an implementation for AddressPrefix.

Reply:
An example: https://play.golang.com/p/z6Gc0YhEQSV